walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -60,50 +52,74 @@
+
+    Collections.sort(busyRanges, TimeRange.ORDER_BY_START);
+
+    return ImmutableList.<TimeRange>builder() 
+              .addAll(busyRanges) 
+              .build();
   }
 
   /** 
-   * Returns a Collection of all free TimeRanges.
-   * Finds free TimeRanges in between busy TimeRanges and takes duration into account.
+   * Returns an ImmutableList of all free TimeRanges.
+   * Finds free TimeRanges in between sorted, busy TimeRanges and takes duration into account.
   */
-  private Collection<TimeRange> findFreeRanges(MeetingRequest request, Collection<TimeRange> busyRanges) {
-    Collection<TimeRange> freeRanges = new ArrayList<TimeRange>();
+  private ImmutableList<TimeRange> findFreeRanges(MeetingRequest request, ImmutableList<TimeRange> sortedBusyRanges) {
+    ArrayList<TimeRange> freeRanges = new ArrayList<>();
 
-    // Default values
     int startFreeRange = TimeRange.START_OF_DAY;
-    int endFreeRange = TimeRange.END_OF_DAY; 
 
-    for (TimeRange busyRange : busyRanges) {
+    for (TimeRange sortedBusyRange : sortedBusyRanges) {
       // Free range ends when a busy range starts
-      endFreeRange = busyRange.start();
+      int endFreeRange = sortedBusyRange.start();
 
       // Make sure we can fit the meeting in the potential range
       if (request.getDuration() <= endFreeRange - startFreeRange){ 
         freeRanges.add(TimeRange.fromStartEnd(startFreeRange, endFreeRange, false));
       }
       
       // Handles over lapping cases
-      startFreeRange = Math.max(startFreeRange, busyRange.end());       
+      startFreeRange = Math.max(startFreeRange, sortedBusyRange.end());       
     } 
 
     // Add the rest of the day, assumming the meeting fits in that time
     if (request.getDuration() < TimeRange.END_OF_DAY - startFreeRange){
       freeRanges.add(TimeRange.fromStartEnd(startFreeRange, TimeRange.END_OF_DAY, true));
     }
 
-    return freeRanges;
+    return ImmutableList.<TimeRange>builder() 
+              .addAll(freeRanges) 
+              .build();
+  }
+
+  /** 
+   * Returns an ImmutableList of all free TimeRanges.
+   * Finds free TimeRanges for both optional and mandatory attendess by considering spaces in between sorted, 
+   * busy TimeRanges and takes duration into account.
+  */
+  private ImmutableList<TimeRange> findFreeRangesAllAttendees(ImmutableList<TimeRange> sortedBusyRangesForMandatoryAttendees, Collection<Event> events, MeetingRequest request) {
+    ImmutableList<TimeRange> sortedBusyRangesForOptionalAttendees = findBusyRanges(request.getOptionalAttendees(), events);
+      
+    ArrayList<TimeRange> busyRangesForAllAttendees = new ArrayList<>();
+    busyRangesForAllAttendees.addAll(sortedBusyRangesForMandatoryAttendees);
+    busyRangesForAllAttendees.addAll(sortedBusyRangesForOptionalAttendees);
+    Collections.sort(busyRangesForAllAttendees, TimeRange.ORDER_BY_START);
+    ImmutableList<TimeRange> immutableBusyRangesForAllAttendees = ImmutableList.<TimeRange>builder() 
```

Comment:
It's not necessary to have immutable in the name since it's obvious from the type declaration.

https://engdoc.corp.google.com/eng/doc/tott/episodes/566.md?cl=head